redis.go

```diff
@@ -1,46 +1,184 @@
 package redis
 
 import (
-	"context"
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
 	"time"
 
+	"github.com/dop251/goja"
 	"github.com/go-redis/redis/v8"
-
 	"go.k6.io/k6/js/common"
 	"go.k6.io/k6/js/modules"
 )
 
 // Register the extension on module initialization, available to
 // import from JS as "k6/x/redis".
 func init() {
-	modules.Register("k6/x/redis", new(Redis))
+	modules.Register("k6/x/redis", New())
 }
 
-// Redis is the k6 extension for a Redis client.
-type Redis struct{}
+type (
+	// RootModule is the global module instance that will create Client
+	// instances for each VU.
+	RootModule struct{}
+
+	// ModuleInstance represents an instance of the JS module.
+	ModuleInstance struct {
+		vu modules.VU
+
+		*Client
+	}
+)
+
+// Ensure the interfaces are implemented correctly
+var (
+	_ modules.Instance = &ModuleInstance{}
+	_ modules.Module   = &RootModule{}
+)
+
+// New returns a pointer to a new RootModule instance
+func New() *RootModule {
+	return &RootModule{}
+}
 
-// Client is the Redis client wrapper.
-type Client struct {
-	client *redis.Client
+// NewModuleInstance implements the modules.Module interface and returns
+// a new instance for each VU.
+func (*RootModule) NewModuleInstance(vu modules.VU) modules.Instance {
+	return &ModuleInstance{vu: vu, Client: &Client{vu: vu}}
 }
 
-// XClient represents the Client constructor (i.e. `new redis.Client()`) and
-// returns a new Redis client object.
-func (r *Redis) XClient(ctxPtr *context.Context, opts *redis.Options) interface{} {
-	rt := common.GetRuntime(*ctxPtr)
-	return common.Bind(rt, &Client{client: redis.NewClient(opts)}, ctxPtr)
+// Exports implements the modules.Instance interface and returns
+// the exports of the JS module.
+func (mi *ModuleInstance) Exports() modules.Exports {
+	return modules.Exports{Named: map[string]interface{}{
+		"Client": mi.NewClient,
+	}}
+}
+
+// NewClient is the JS constructor for the redis Client.
+//
+// Under the hood, the redis.UniversalClient will be used. The universal client
+// supports failover/sentinel, cluster and single-node modes. Depending on the options,
+// the internal universal client instance will be one of those.
+//
+// The type of the underlying client depends on the following conditions:
+// 1. If the MasterName option is specified, a sentinel-backed FailoverClient is used.
+// 2. if the number of Addrs is two or more, a ClusterClient is used.
+// 3. Otherwise, a single-node Client is used.
+func (mi *ModuleInstance) NewClient(call goja.ConstructorCall) *goja.Object {
+	rt := mi.vu.Runtime()
+
+	var optionsArg map[string]interface{}
+	err := rt.ExportTo(call.Arguments[0], &optionsArg)
+	if err != nil {
+		common.Throw(rt, errors.New("unable to parse options object"))
+	}
+
+	opts, err := newOptionsFrom(optionsArg)
+	if err != nil {
+		common.Throw(rt, fmt.Errorf("invalid options; reason: %w", err))
+	}
+
+	client := &Client{
+		vu: mi.vu,
+		redisClient: redis.NewUniversalClient(&redis.UniversalOptions{
```

Comment:
I would argue you need to set the Dialer and TLSConfig here to the ones from the lib.State

Reply:
Indeed 🙇🏻